### DISCUSSION TITLE
Some "orignal" videos are been exported as xml.

### DISCUSSION THREAD

**Discussion by chengguangnan** [RhetTbull/osxphotos#1379] [General]
I found that the _orignal_ videos were exported as a xml file with .mov extention, meanwhile _edited_ videos were exported correctly.

I'm assuming that this only happen to videos?

<img width="692" alt="Screenshot 2024-01-12 at 12 18 34 AM" src="https://github.com/RhetTbull/osxphotos/assets/861069/e3e2584d-7372-4cfc-80b5-e653909017c8">

The content of that xml file `IMG_4990.mov` is 

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>adjustmentBaseVersion</key>
	<integer>0</integer>
	<key>adjustmentData</key>
	<data>
	....
	</data>
	<key>adjustmentFormatIdentifier</key>
	<string>com.apple.photo</string>
	<key>adjustmentFormatVersion</key>
	<string>1.6</string>
	<key>adjustmentRenderTypes</key>
	<integer>0</integer>
</dict>
</plist>
```

**Comment by RhetTbull** [comment]
I am not able to replicate this.  Please run the following command and send the output to osxphotos.py@gmail.com

`osxphotos debug-dump --dump photos --name IMG_4990 > debug.log`

**Reply by chengguangnan** [reply]
Sent.

**Comment by chengguangnan** [comment]
After a few hours of testing, it seems to me that, this mostly happens to videos that were been "trimmed" and saved on the orignal file, and the orignal is often `missing` upon the first export.

The first export (with --download-missing) of a such file will produce a xml (plist?) file with .mov extention, but the next fresh export (at a different folder) would export the actual video content. 

My testing commands:
```
cd debug-5157
osxphotos debug-dump --dump photos --uuid A8144734-26B6-48D5-966E-21A2A3996B49 > first-dump.log
osxphotos --debug export . --uuid A8144734-26B6-48D5-966E-21A2A3996B49 --download-missing --report report.csv
osxphotos debug-dump --dump photos --uuid A8144734-26B6-48D5-966E-21A2A3996B49 > second-dump.log

cd..
mkdir debug-5157-2
cd debug-5157-2/
osxphotos --debug export . --uuid A8144734-26B6-48D5-966E-21A2A3996B49 --download-missing --report report.csv
osxphotos debug-dump --dump photos --uuid A8144734-26B6-48D5-966E-21A2A3996B49 > third-dump.log
```

A `vimdiff` on those debug-dumps suggest that the major differences before and after the first export were `localAvailability`, `isMissing`, `path`, etc.  (BTW, I did a third dump after two successful exports and `localAvailability`, `isMissing`, `path` went back to negative, but I guess this is irrelevant here)


The exported file size are:
<img width="321" alt="Screenshot 2024-01-12 at 4 01 13 PM" src="https://github.com/RhetTbull/osxphotos/assets/861069/6168df31-66b1-4cd4-8583-b3e15ad98f8e">


### Summary

I'm assuming this issue related to videos that were been "trimed and saved on orignal". I'm looking into the code, right now I'm at `export_photo_to_directory`, hopeful I would add some breakpoint to inspect and dig out more infomation.

**Reply by RhetTbull** [reply]
>  I'm looking into the code, right now I'm at export_photo_to_directory, hopeful I would add some breakpoint to inspect 

If you're comfortable with the debugger you can invoke it directly from the osxphotos command line using a hidden `--breakpoint` option. This option must be passed *before* the `export` command:

`osxphotos --breakpoint osxphotos.photoexporter::PhotoExporter.export export ...rest of your export options`

In this case, breakpoint is on PhotoExporter().export() which is where the export actually happens. The syntac is `file.path::function` or `file.path::class.method`

The breakpoint will be right before the function you want to inspect so you'll need to step once to enter the export() function.

**Comment by RhetTbull** [comment]
I did some more digging with the built in osxphotos `snap` and `diff` commands as well as my [dirsnapshot](https://github.com/RhetTbull/dirsnapshot) tool to see what gets changed when a video is trimmed. On my macBook (Ventura 13.5.1) with "Optimize Mac Storage" enabled, the video gets downloaded when viewed, then when trimmed the plist file with adjustments data gets created/modified and a new preview image is crated. The plist file is at `Photos Library.photoslibrary/resources/renders/C/C3388AEA-EFB6-46F8-A20D-E91688DD059D.plist` and is an XML property list file with the ajdustment data that matches what you are seeing as the XML file that got exported.  There is no derivative (edited version) of the video created. Photos must look at the adjustment data in the plist and show the trimmed video when asked to play it.

When I export this with osxphotos, osxphotos exports only the original video and doesn't even report that the editing photo is missing, which it should:

```
Exporting 1 photo to /Users/rhet/Desktop/export...
Exporting IMG_0370.MOV (C3388AEA-EFB6-46F8-A20D-E91688DD059D.mov) (1/1)
Exported IMG_0370.MOV to /Users/rhet/Desktop/export/IMG_0370.MOV
Exported /Users/rhet/Desktop/export/IMG_0370.MOV
Processed: 1 photo, exported: 1, missing: 0, error: 0
Elapsed time: 0:00:00
```

The XML plist file you got is not exported nor is an edited version of the video.

So there may be more than one bug here. I'll do some more debugging to see what I can figure out.

**Reply by RhetTbull** [reply]
Ok, I figured out why the edited version was skipped. The video I chose to test with was a slow-motion video and these always show hasadjustments = True even if not edited so the code skips these if there's no edited version:

```
    # slow_mo photos will always have hasadjustments=True even if not edited
    if photo.hasadjustments and photo.path_edited is None:
        if photo.slow_mo:
            export_original = True
            export_edited = False
```

I'll test again with a regular video.

**Reply by RhetTbull** [reply]
I found another bug (#1380) but don't think it's related and the results for a regular video that's been trimmed are the same. I still don't see an edited version of the video and the XML does not get exported as the original version.

**Comment by RhetTbull** [comment]
I've done quite a bit more experimentation with editing videos and exporting and I cannot reproduce the behavior you're seeing.  Will keep working on this as time permits. Nothing in the log files you sent jumps out.

**Comment by chengguangnan** [comment]
I avoided the problem by using `--photokit`. 

Regards to Apple Script exporting, upon a 1000 video export, about 20 will be exported as a small plist XML file upon the first attempt. 

I looked into the code at https://github.com/RhetTbull/PhotoScript/blob/ffe1b0f892c1e8ab788ffe5a1467e341459292ed/photoscript/__init__.py#L566-L569, it seems that there is a filename rename logic, it seems to me that it's likely somehow the Apple Script exported the plist file and it's been mistakenly renamed to .MOV? Maybe we could add some log here or do not delete the TemporaryDirectory to figure out what happened?

**Reply by RhetTbull** [reply]
This is helpful -- I'll check the PhotoScript code to see if this can be detected. PhotoScript calls the Photos AppleScript API to export the photos/videos so if this is happening, it's because Photos is doing it. If so, I'll file a bug report with Apple.

**Reply by RhetTbull** [reply]
I also created a new issue in PhotoScript for this.

**Reply by RhetTbull** [reply]
I have tried to replicate this in PhotoScript but so far, using a library with "optimize Mac storage" so files are not local, have not been able to. The .MOV is always properly exported.
